src/java.base/share/classes/jdk/internal/classfile/impl/ClassHierarchyImpl.java

```diff
@@ -101,66 +108,90 @@ public boolean isAssignableFrom(ClassDesc thisClass, ClassDesc fromClass) {
     }
 
     public static final class CachedClassHierarchyResolver implements ClassHierarchyResolver {
-
-        //this instance should never appear in the cache nor leak out
+        //this instance should leak out, appears only in cache in order to utilize Map.computeIfAbsent
         private static final ClassHierarchyResolver.ClassHierarchyInfo NOPE =
                 new ClassHierarchyResolver.ClassHierarchyInfo(null, false, null);
 
+        private final ClassHierarchyResolver delegate;
+        private final Map<ClassDesc, ClassHierarchyInfo> resolvedCache;
+
+        public CachedClassHierarchyResolver(ClassHierarchyResolver delegate, Map<ClassDesc, ClassHierarchyInfo> resolvedCache) {
+            this.delegate = delegate;
+            this.resolvedCache = resolvedCache;
+        }
+
+        @Override
+        public ClassHierarchyInfo getClassInfo(ClassDesc classDesc) {
+            var ret = resolvedCache.computeIfAbsent(classDesc, new Function<>() {
+                @Override
+                public ClassHierarchyInfo apply(ClassDesc classDesc) {
+                    var ret = delegate.getClassInfo(classDesc);
+                    return ret == null ? NOPE : ret;
+                }
+            });
+            return ret == NOPE ? null : ret;
+        }
+    }
+
+    public static final class ResourceParsingClassHierarchyResolver implements ClassHierarchyResolver {
+        public static final Function<ClassDesc, InputStream> SYSTEM_STREAM_PROVIDER = new Function<>() {
+            @SuppressWarnings("removal")
+            @Override
+            public InputStream apply(ClassDesc cd) {
+                var sm = System.getSecurityManager();
+                if (sm != null) {
+                    return AccessController.doPrivileged(new PrivilegedAction<>() {
```

Comment:
This is out of my competency, need another reviewer for use of `PriviledgedAction` in this context.